apps/src/Sounds.js

```diff
@@ -328,11 +378,76 @@ Sounds.prototype.playURL = function(url, playbackOptions) {
   }
 };
 
+/**
+ * @param {Promise} promise the promise that will resolve to the ArrayBuffer bytes to be played
+ * @param {object} playbackOptions config for the playing of the sound.
+ * @param {object} cacheParams optional parameters for caching the result of promise
+ */
+Sounds.prototype.addPromiseToSpeechQueue = function(
+  promise,
+  playbackOptions,
+  cacheParams = null
+) {
+  this.speechQueue.push({
+    promise: promise,
+    playbackOptions: playbackOptions,
+    cacheParams: cacheParams
+  });
+  this.checkSpeechQueue();
+};
+
+Sounds.prototype.checkSpeechQueue = async function() {
+  if (!this.speechPlaying && this.speechQueue.length > 0) {
+    this.speechPlaying = true;
+    let nextSpeech = this.speechQueue.shift();
+    let bytes = await nextSpeech['promise'];
+    if (bytes === null) {
+      if (
+        this.textBytesByLanguage[nextSpeech['cacheParams']['language']][
+          nextSpeech['cacheParams']['text']
+        ]['hasProfanity']
+      ) {
+        nextSpeech['cacheParams']['profanityFoundCallback'](
+          this.textBytesByLanguage[nextSpeech['cacheParams']['language']][
+            nextSpeech['cacheParams']['text']
+          ]['profaneWords']
+        );
+        this.onSpeechFinished();
+      } else {
+        bytes = this.textBytesByLanguage[nextSpeech['cacheParams']['language']][
+          nextSpeech['cacheParams']['text']
+        ][nextSpeech['cacheParams']['gender']];
+        this.playSpeechBytes(bytes.slice(0), nextSpeech['playbackOptions']);
+      }
+    } else if (bytes === 'profanity found') {
+      this.onSpeechFinished();
+    } else if (nextSpeech['cacheParams'] !== null) {
+      const bytes = await nextSpeech['promise'];
+      this.registerTextBytes(
+        nextSpeech['cacheParams']['language'],
+        nextSpeech['cacheParams']['text'],
+        nextSpeech['cacheParams']['hasProfanity'],
+        nextSpeech['cacheParams']['profaneWords'],
+        nextSpeech['cacheParams']['gender'],
+        bytes.slice(0)
+      );
+      this.playSpeechBytes(bytes, nextSpeech['playbackOptions']);
+    } else {
+      this.playSpeechBytes(bytes.slice(0), nextSpeech['playbackOptions']);
+    }
+  }
+};
+
+Sounds.prototype.onSpeechFinished = function() {
+  this.speechPlaying = false;
+  this.checkSpeechQueue();
+};
+
 /**
  * @param {ArrayBuffer} bytes of the sound to play.
  * @param {object} playbackOptions config for the playing of the sound.
  */
-Sounds.prototype.playBytes = function(bytes, playbackOptions) {
+Sounds.prototype.playSpeechBytes = function(bytes, playbackOptions) {
```

Comment:
Can you double check if this function is used anywhere else that would need to be updated?

Reply:
oh right I forgot we meant to make it reusable, I'll switch it back